### PR TITLE
Unnamed construct analysis

### PR DESCRIPTION
**Pull request by houli**
Unnamed compound constructs sent in API response
No backing UI yet

### CODE REVIEW

File: panacea/lib/panacea/pml/analysis/unnamed.ex [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        [%{type: type, line: Keyword.get(attrs, :line)} | result]
+      else
+        result
+      end
+
+    Enum.reduce(children, result, fn (child, acc) ->
+      analyse(acc, child)
+    end)
+  end
+  defp analyse(result, {_, _, children}) when is_list(children) do
```
Comment (c-brenn) [inline panacea/lib/panacea/pml/analysis/unnamed.ex]: are there any constructs that have a list of children that are not covered by the composite case?

Comment (houli) [reply panacea/lib/panacea/pml/analysis/unnamed.ex]: I don't think so, I was just being safe

Comment (houli) [reply panacea/lib/panacea/pml/analysis/unnamed.ex]: Oh actually, `action` yes now that I think about it and process

Comment (c-brenn) [reply panacea/lib/panacea/pml/analysis/unnamed.ex]: unnamed `action` and `process` will cause a syntax error

Comment (houli) [reply panacea/lib/panacea/pml/analysis/unnamed.ex]: Yeah, we still want to process them. If I didn't have this case they would never be processed further

Comment (c-brenn) [reply panacea/lib/panacea/pml/analysis/unnamed.ex]: Right, but I think we can put them in the composite list and get the same effect?

**Review by c-brenn — COMMENTED** [review]
Do you see much benefit in doing the analysis in two separate walks of the AST? Looks like we could get both in one without much hassle.

The unlikely possibility is that as we add more analysis, multiple iterations may not be performant.

**Comment by houli** [comment]
It's very fast as it is right now, I don't think it's much of a concern.
The main reason I split it up was to make it more readable and to separate out the two different analyses. Handling both cases at the same time is liable to make the code quite messy.

File: panacea/lib/panacea/pml/analysis/unnamed.ex (hunk 1 above)
Comment (houli) [reply panacea/lib/panacea/pml/analysis/unnamed.ex]: Changed

**Comment by c-brenn** [comment]
Agreed, it could be messy. Instead of re-analysing the entire AST in each walk, we could re-analyse some nodes in one walk and keep the code clean. Rough idea below.

Up to you which you think is cleaner.

```elixir
  defp analyse(result, {type, attrs, children}) when type in @composite do
    result =
      result
      |> Naming.ensure_named(attrs)
      |> Naming.check_name_clash(attrs)
      |> Deprecations.check_task_usage(type)

    Enum.reduce(children, result, fn (child, acc) ->
      analyse(acc, child)
    end)
  end
  defp analyse(result, {_, _, child}) when is_tuple(child) do
    analyse(result, child)
  end
  defp analyse(result, value_node) do
    Drugs.store_if_drug(result, value_node)
  end
```

**Review by c-brenn — APPROVED** [review]

**Comment by houli** [comment]
I think the way we have things at the minute is quite straight-forward to add more code even if you're not familiar, just define another module under analysis and write an analyse function, call it in the main analysis module with the AST. Don't want to have to sprinkle bits around lots of other modules

**Comment by c-brenn** [comment]
Do we want the UI folk to do their (they're there) bits on this PR or separate?

**Comment by houli** [comment]
We can do it separately, I'll merge this in now

**Pull request merged**